include/pybind11/pybind11.h

```diff
@@ -1324,6 +1325,10 @@ class class_ : public detail::generic_type {
                 return nullptr;
             return new buffer_info(((capture *) ptr)->func(caster));
         }, ptr);
+        weakref(m_ptr, cpp_function([ptr](handle wr) {
+            delete ptr;
+            wr.dec_ref();
+        })).release();
```

Comment:
I'm thinking we could still refactor this to something like

```
template <typename F>
void cleanup_through_weakref(handle parent, F &&f) {
    weakref(parent, cpp_function([f=std::forward<F>(f)](handle wr) {
        f();
        wr.dec_ref();
    })).release();
};
```

But we'd also need some workaround for C++11's inability to capture-by-move. So let's maybe leave this for a next PR, if this pattern keeps coming back.